File: BOJ/수열/15654-N과 M(5).java
import java.io.*;
import java.util.*;
public class Main {
    static int n, m;
    static int[] num;
    static ArrayList<Integer> ans;
    static int[] visited;
    static void backtracking() {
        if (ans.size() == m) {
            for (int n : ans) {
                System.out.print(n + " ");
            }
            System.out.println();
            return;
        }

        for (int i = 0; i < n; i++) {
            if (visited[i] == 0) {
                visited[i] = 1;
                ans.add(num[i]);
                backtracking();
                ans.remove(ans.size() - 1);
                visited[i] = 0;
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        ans = new ArrayList<>();
        visited = new int[n];
        num = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            num[i] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(num);

        backtracking();
    }
}

File: BOJ/수열/15654-N과 M(5).cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;


int n, m;
vector<int> visited;
vector<int> ans;
vector<int> arr;
void backtracking(int idx) {

    if (ans.size() == m) {
        for (auto num : ans) {
            cout << num << " ";
        }
        cout << "\n";
        return;
    }

    for (int i = 0; i < n; i++) {
        if (visited[i] == 0) {
            visited[i] = 1;
            ans.push_back(arr[i]);
            backtracking(idx + 1);
            ans.pop_back();
            visited[i] = 0;
        }
    }

}

int main() {
    cin >> n >> m;
    arr.assign(n, 0);
    visited.assign(n, 0);
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
    }
    sort(arr.begin(), arr.end());
    backtracking(0);
    return 0;
}